deepchem/feat/vocabulary_builders/grover_vocab.py

```diff
@@ -0,0 +1,414 @@
+        """Loads vocabulary from the specified json file
+
+        Parameters
+        ----------
+        fname: str
+            JSON file containing vocabulary
+
+        Returns
+        -------
+        vocab: GroverAtomVocabularyBuilder
+            A grover atom vocabulary builder which can be used for encoding
+        """
+        with open(fname, 'r') as f:
+            data = json.load(f)
+        vocab = cls()
+        vocab.stoi, vocab.itos = data['stoi'], data['itos']
+        return vocab
+
+    @staticmethod
+    def atom_to_vocab(mol: RDKitMol, atom: RDKitAtom) -> str:
+        """Convert atom to vocabulary.
+
+        Parameters
+        ----------
+        mol: RDKitMol
+            an molecule object
+        atom: RDKitAtom
+            the target atom.
+
+        Returns
+        -------
+        vocab: str
+            The generated atom vocabulary with its contexts.
+
+        Example
+        -------
+        >>> from rdkit import Chem
+        >>> mol = Chem.MolFromSmiles('[C@@H](C)C(=O)O')
+        >>> GroverAtomVocabularyBuilder.atom_to_vocab(mol, mol.GetAtomWithIdx(0))
+        'C_C-SINGLE2'
+        >>> GroverAtomVocabularyBuilder.atom_to_vocab(mol, mol.GetAtomWithIdx(3))
+        'O_C-DOUBLE1'
+        """
+        nei: Dict[str, int] = Counter()
+        for a in atom.GetNeighbors():
+            bond = mol.GetBondBetweenAtoms(atom.GetIdx(), a.GetIdx())
+            nei[str(a.GetSymbol()) + "-" + str(bond.GetBondType())] += 1
+        keys = list(nei.keys())
+        # sorting the atoms neighbors
+        keys.sort()
+        output = atom.GetSymbol()
+        # concatenating the sorted neighbors
+        for key in keys:
+            output = "%s_%s%d" % (output, key, nei[key])
+        return output
+
+    def _make_reverse_mapping(self, itos):
+        return {tok: i for i, tok in enumerate(itos)}
+
+    def encode(self, mol: RDKitMol, atom: RDKitAtom) -> str:
+        """Encodes an atom in a molecule
+
+        Parameter
+        ---------
+        mol: RDKitMol
+            An RDKitMol object
+        atom: RDKitAtom
+            An atom in the molecule
+
+        Returns
+        -------
+        vocab: str
+            The vocabulary of the atom in the molecule.
+        """
+        return self.stoi.get(self.atom_to_vocab(mol, atom))
+
+
+class GroverBondVocabularyBuilder(VocabularyBuilder):
+    """Bond Vocabulary Builder for Grover
+
+    This module can be used to generate bond vocabulary from SMILES strings for
+    the GROVER pretraining task. For each bond in a molecule, the vocabulary context is the
+    node-edge-count of the bond where node is the neighboring atom, edge is the type of bond
+    and count is the number of such node-edge pairs for the bond in its one hop
+    neighborhood. The algorithm enumerates vocabulary of all bonds
+    in the dataset and makes a vocabulary to index mapping by sorting the vocabulary
+    by frequency and then alphabetically.
```

Comment:
Can you add a detailed example here like the one you have above for the atom vocab builder?

Reply:
added example